lib/schema-controller.js

```diff
@@ -72,13 +72,54 @@ class SchemaController {
     return this.schemaBucket.getSchemas()
   }
 
-  // Schema Controller compilers holder
+  // Called when compilers are recreated from builders, but the compiler
+  // was overridden via the legacy functions 'setValidatorCompiler' or
+  // 'setSerializerCompiler.'  It exists as a separate function only
+  // because c8 flags an inline arrow function as uncovered.
+  wrapAsBuilder (compiler) {
+    return compiler
+  }
+
   setValidatorCompiler (validatorCompiler) {
+    // Set up as if the fixed validator compiler had been provided
+    // by a custom 'options.compilersFactory.buildValidator' that
+    // always returns the same compiler object. This is required because:
+    //
+    // - setValidatorCompiler must immediately install a compiler to preserve
+    //   legacy behavior
+    // - setupValidator will recreate compilers from builders in some
+    //   circumstances, so we have to install this adapter to make it
+    //   behave the same if the legacy API is used
+    //
+    // The cloning of the compilersFactory object is necessary because
+    // we are aliasing the parent compilersFactory if none was provided
+    // to us (see constructor.)
+    this.compilersFactory = Object.assign(
+      {},
+      this.compilersFactory,
+      { buildValidator: this.wrapAsBuilder.bind(this, validatorCompiler) })
     this.validatorCompiler = validatorCompiler
     this.isCustomValidatorCompiler = true
   }
 
   setSerializerCompiler (serializerCompiler) {
+    // Set up as if the fixed serializer compiler had been provided
+    // by a custom 'options.compilersFactory.buildSerializer' that
+    // always returns the same compiler object. This is required because:
+    //
+    // - setSerializerCompiler must immediately install a compiler to preserve
+    //   legacy behavior
+    // - setupSerializer will recreate compilers from builders in some
+    //   circumstances, so we have to install this adapter to make it
+    //   behave the same if the legacy API is used
+    //
+    // The cloning of the compilersFactory object is necessary because
+    // we are aliasing the parent compilersFactory if none was provided
+    // to us (see constructor.)
+    this.compilersFactory = Object.assign(
+      {},
+      this.compilersFactory,
```

Comment:
```suggestion
    Object.assign(
      this.compilersFactory,
```

Reply:
It is also written 😂 

Anyway: could you add a test named "the setValidatorCompiler overwrites the compilersFactory validator"?

I think we don't have this use case.

Then LGTM 👍🏼

Reply:
@Eomm I wrote some tests.  Can you please take a look?   Do we agree the last one should succeed?  (because it fails currently)

In the "backwards" case where the child instance has a custom builder, it does not work.  It crashes out throwing that the extended schema is invalid, meaning https://github.com/fastify/fastify/blob/f2b165c45b0f70203a9ccc70b2b636a18990b8f2/test/schema-validation.test.js#L1167 did nothing.

Reply:
PS: I changed the style of the tests a bit to actually USE custom schema instead of just testing that the compilers get called, because I felt at least some of the tests should make sure the resulting compiled schemas are used at the right times.

Reply:
update: test case being discussed here is temporarily commented out

Reply:
> Can you please take a look? Do we agree the last one should succeed? (because it fails currently)

Yes and no 😄 

The issue is that:

- `setValidatorCompiler` is sync in the sense that it sets `this.validatorCompiler` when it runs
- meanwhile `setSchemaController` is totally async and related to the applicationtion loadings, so I think they will fight for sure

To fix the test, I moved the `setValidatorCompiler` after the plugin registration, this ensures all the events are processed in order and the test passes.

I don't think we have a golden bullet here to make it work in a different way, because we can't execute `buildValidator` logic at that time, because the user may run an `addSchema` later!
It is a chicken and egg issue this one.

Reply:
> To fix the test, I moved the `setValidatorCompiler` after the plugin registration, this ensures all the events are processed in order and the test passes.
> 
> I don't think we have a golden bullet here to make it work in a different way, because we can't execute `buildValidator` logic at that time, because the user may run an `addSchema` later! It is a chicken and egg issue this one.

Yeah I also suspected it would be questionable what "correct" behavior would be.  One thing I was considering is that the user's code order should decide.  In other words, when you call `setSchemaController`, it should synchronously `delete` the installed custom compilers, so that they will be asynchronously recreated on the next `setupValidator` and `setupSerializer`.   In other words, they would behave as if they synchronously compete, so the behavior matches the user's code order?

I just wasn't sure if that would work, so I figured we should discuss.